### DISCUSSION TITLE
Simplify service API

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12339] [Ideas]
## Problem

The Service API is powerful, but confusing - especially to beginners.

- `Service.start()` starts a service, but you're not supposed to use it, except in a niche use case.
- `Service.up()` also starts a service, but differently - only to expose ports to the client's network, and should only be called in the CLI. If you call from a Dagger Function, it will expose ports on the function container's network.
- `Service.stop()` stops a service, but you're not supposed to use it, except in a niche use case.
- `Container.asService()` converts a container to a service. Makes sense in code, but confusing to call in the CLI (see #7399 )

I am not challenging the engineering and design decisions that led to this. But simply observing the current user experience, which can be confusing.

## Solution

Adjust the design to minimize user confusion, with as few breaking changes as possible.

## User feedback

- [what triggers stopping a service declared under a container with WithServiceBinding?](https://discord.com/channels/707636530424053791/1123687185837740053/1273365938351374436)
- 
## Design proposals

Please add ideas in the comments. I will incorporate them here.

**Comment by shykes** [comment]
In #7199 we introduce the concept of a *Context API*. One possible application of the context API is interacting with the client's networking context. This would take over and expand networking functions currently in `dag.Host()`.

What if we leveraged this context API to allow `Service.up()` to work from inside the function? You could pass an optional context argument, to specify *where* you want to bring the service up. (kudos to @levlaz for this idea).

For example:

```golang
// Bring the service up in our client's context (eg. in the host)
svc.Up(dag.Context().Network())
// Current default: bring the service up in a sandboxed context inside the container
svc.Up()
```

Of course this would require a trust model. But so would the rest of the context API proposal (for example we are discussing giving access to the client's credentials)

**Comment by levlaz** [comment]
I would love to see this, if we can find a way to do this safely it
would make the DX feel a lot smoother.

**Comment by shykes** [comment]
> ```go
> // Bring the service up in our client's context (eg. in the host)
> svc.Up(dag.Context().Network())
> // Current default: bring the service up in a sandboxed context inside the container
> svc.Up()
> ```

Follow-up questions:

- Would this allow merging `Service.start()` and `Service.up()`?
- What exactly would be the type passed as argument to `up()`? What's the API for abstracting away a service endpoint implementation?

**Comment by jedevc** [comment]
See some other linked issues as well (just for completeness):
- https://github.com/dagger/dagger/issues/7391
- https://github.com/dagger/dagger/issues/6553 
- https://github.com/dagger/dagger/issues/6596

---

I agree somewhat on the Service `Start`/`Stop` point. I think manual access should still be permitted (e.g. to force stop a service with a specific signal), but we should have better defaults - I've imagined a `svc.WithStop(dagger.SIGTERM)` or similar to allow manually setting the signal by which a service should be stopped.

My use case is that I would like the opportunity to stop a test service with `SIGQUIT` iff the tests time out, but not in the general case.

---

Agreed very much on the need to improve `up` - it's very difficult to call from code (since the underlying tunnel can't be closed).

**Comment by sipsma** [comment]
There's more APIs that weren't listed in the description:
* `dag.Host().Tunnel()` - serves a containerized service endpoint on the host
   * Also worth noting that `.Up` is just a chainable version of `Tunnel` essentially, allowing it to be called in a single CLI command.
* `dag.Host().Service()` - turns a host address+port into a service that can be used by containers

**Comment by jedevc** [comment]
Linking in other related issues:
- https://github.com/dagger/dagger/issues/8037 
- https://github.com/dagger/dagger/issues/8046

**Comment by shykes** [comment]
Added relevant user feedback in the top comment

**Comment by sagikazarmark** [comment]
Would be nice to allow services to talk to each other:

```mermaid
flowchart LR
    ServiceA --> ServiceB
    ServiceB --> ServiceA
```

**Comment by Spenhouet** [comment]
I'm confused. In my python code I'm trying to serve an API via:

```python
await container.as_service(args=cmd).up(ports=[PortForward(port, port)])
```

The service starts up and display that is setup tunneling but the API is not reachable.

When I just return 

```python
container.as_service(args=cmd)
```

and then in CLI add a 

```
up --ports 8005:8005
```

then it's reachable (with the same tunneling message).
